File: frontend/lib/api/notifications.ts
import { apiFetch } from "@/lib/api/client";
import type { Notification, NotificationList } from "@/types/api";

export function getNotifications() {
  return apiFetch<NotificationList>("/notifications");
}

export function markNotificationRead(id: string) {
  return apiFetch<Notification>(`/notifications/${id}/read`, { method: "POST" });
}

export function markAllNotificationsRead() {
  return apiFetch<void>("/notifications/read-all", { method: "POST" });
}

File: backend/app/routers/notifications.py
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.notification import NotificationList, NotificationOut
from app.services import notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=NotificationList)
async def list_notifications(
    user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> NotificationList:
    items, unread_count = await notification_service.list_notifications(db, user)
    return NotificationList(items=[NotificationOut.model_validate(n) for n in items], unread_count=unread_count)


@router.post("/{notification_id}/read", response_model=NotificationOut)
async def mark_notification_read(
    notification_id: uuid.UUID, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> NotificationOut:
    notification = await notification_service.mark_read(db, notification_id, user)
    return NotificationOut.model_validate(notification)


@router.post("/read-all", status_code=status.HTTP_204_NO_CONTENT)
async def mark_all_read(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)) -> None:
    await notification_service.mark_all_read(db, user)
